contracts/src/v0.8/ccip/pools/SiloedLockReleaseTokenPool.sol

```diff
@@ -0,0 +1,286 @@
+    ChainSiloConfig memory remoteConfig = s_siloedChainConfigs[remoteChainSelector];
+
+    if (!remoteConfig.isSiloed) revert ChainNotSiloed(remoteChainSelector);
+
+    address oldRebalancer = remoteConfig.rebalancer;
+
+    s_siloedChainConfigs[remoteChainSelector].rebalancer = newRebalancer;
+
+    emit SiloedChainRebalancerSet(remoteChainSelector, newRebalancer, oldRebalancer);
+  }
+
+  /// @notice Sets the Rebalancer address for unsiloed chains.
+  /// @dev Only callable by the owner.
+  /// @param newRebalancer the address allowed to add liquidity for the given siloed chain.
+  function setUnsiloedChainRebalancer(
+    address newRebalancer
+  ) external onlyOwner {
+    address oldRebalancer = s_unsiloedChainRebalancer;
+
+    s_unsiloedChainRebalancer = newRebalancer;
+
+    emit UnsiloedChainRebalancerSet(newRebalancer, oldRebalancer);
+  }
+
+  /// @notice Adds liquidity to the pool. The tokens should be approved first.
+  /// @param remoteChainSelector the remote chain to set. If the chain is not siloed, the liquidity will be shared among all
+  /// non-siloed chains.
+  /// @param amount The amount of liquidity to provide.
+  /// @dev Only the rebalancer for the chain can add liquidity
+  function provideSiloedLiquidity(uint64 remoteChainSelector, uint256 amount) external {
+    _provideLiquidity(remoteChainSelector, amount);
+  }
+
+  /// @notice Adds liquidity to the pool for unsiloed chains. Function is used to support legacy liquidity operations
+  /// by using a function selector available to previous L/R pools.
+  /// @dev Since the remoteChainSelector 0 should never be applied to a real chain, it is used to designate unsiloed chains.
+  /// @param amount The amount of liquidity to provide.
+  function provideLiquidity(
+    uint256 amount
+  ) external {
+    _provideLiquidity(0, amount);
+  }
+
+  function _provideLiquidity(uint64 remoteChainSelector, uint256 amount) internal {
+    if (msg.sender != getRebalancerByChain(remoteChainSelector)) revert Unauthorized(msg.sender);
+
+    // Storage is used instead of memory to save gas, as the state may need to be updated if the chain is siloed.
+    ChainSiloConfig storage remoteConfig = s_siloedChainConfigs[remoteChainSelector];
+
+    if (remoteConfig.isSiloed) {
+      remoteConfig.siloedBalance += amount;
+    } else {
+      s_unsiloedTokenBalance += amount;
+    }
+
+    i_token.safeTransferFrom(msg.sender, address(this), amount);
+    emit LiquidityAdded(remoteChainSelector, msg.sender, amount);
+  }
+
+  /// @notice Removes liquidity from the pool for unsiloed chains. Function is used to support legacy liquidity operations
+  /// by using a function selector available to previous L/R pools.
+  /// @dev Since the remoteChainSelector 0 should never be applied to a real chain, it is used to designate unsiloed chains.
+  /// @param amount The amount of liquidity to remove.
+  function withdrawLiquidity(
```

Comment:
Maybe just merge and have 0 be a value the user submits and we don't have a separate func?

Reply:
It was requested that this remain for the finance team to not have to do more tooling work by using the existing function selectors.